docs/coding-guidelines.md

```diff
@@ -0,0 +1,269 @@
+        - [Prioritize local work over new work from a remote](#prioritize-local-work-over-new-work-from-a-remote)
+    - [Bound everything](#bound-everything)
+        - [Channels](#channels)
+        - [Local queues](#local-queues)
+        - [Further reading](#further-reading)
+    - [No premature optimizations](#no-premature-optimizations)
+    - [Keep things sequential unless proven to be slow](#keep-things-sequential-unless-proven-to-be-slow)
+    - [Use `async/await` for sequential execution only](#use-asyncawait-for-sequential-execution-only)
+    - [Don't communicate by sharing memory; share memory by communicating.](#dont-communicate-by-sharing-memory-share-memory-by-communicating)
+        - [Further Reading](#further-reading)
+
+<!-- markdown-toc end -->
+
+
+Below is a set of coding guidelines followed across the rust-libp2p code base.
+
+## Hierarchical State Machines
+
+If you sqint, rust-libp2p is just a big hierarchy of [state
+machines](https://en.wikipedia.org/wiki/Finite-state_machine) where parents pass
+events down to their children and children pass events up to their parents.
+
+<details>
+    <summary>Code</summary>
+
+    ```
+    @startuml
+    Swarm <|-- RootBehaviour
+    Swarm <|-- ConnectionPool
+    Swarm <|-- Transport
+    RootBehaviour <|-- PingBehaviour
+    RootBehaviour <|-- IdentifyBehaviour
+    RootBehaviour <|-- KademliaBehaviour
+
+    Swarm : poll()
+    RootBehaviour : poll()
+    ConnectionPool : poll()
+    Transport : poll()
+    PingBehaviour : poll()
+    IdentifyBehaviour : poll()
+    KademliaBehaviour : poll()
+    @enduml
+    ```
+</details>
+
+```
+                                         ,------.
+                                         |Swarm |
+                                         |------|
+                                         |poll()|
+                                         `------'
+                                             |
+                                             |
+                   ,-------------.   ,--------------.   ,---------.
+                   |RootBehaviour|   |ConnectionPool|   |Transport|
+                   |-------------|   |--------------|   |---------|
+                   |poll()       |   |poll()        |   |poll()   |
+                   `-------------'   `--------------'   `---------'
+                           |
+,-------------.  ,-----------------.   ,-----------------.
+|PingBehaviour|  |IdentifyBehaviour|   |KademliaBehaviour|
+|-------------|  |-----------------|   |-----------------|
+|poll()       |  |poll()           |   |poll()           |
+`-------------'  `-----------------'   `-----------------'
+```
+
+Using hierarchical state machines is a deliberate choice throughout the
+rust-libp2p code base. It makes reasoning about control and data flow easy and
+works well with Rust's `Future` model. The archetecture pattern of
+hierarchical state machines should be used wherever possible.
```

Comment:
Are there downsides to this design? If so, should we list them?

Reply:
In my experience, the main downsides are:

- It _feels_ more verbose but I think it is not actually in reality when you would compare two solutions side by side.
- The mix of control flow (`loop`, `return`, `break`, `continue`) together with the asynchronous and thus decoupled communication via events can be very hard to understand.

Both are a form of complexity that we are trading for correctness and performance which seems typical in Rust land.

Reply:
I documented the above in https://github.com/libp2p/rust-libp2p/pull/2780/commits/0ef322a27252bd53d87205419969fcdcefdceade. @thomaseizinger I really liked your wording which I mostly adopted. Hope you don't mind.

@melekes in case you can think of other downsides, please comment and let's get them included in the discussion and the document.